py4web/utils/auth_plugins/oauth2google_scoped.py

```diff
@@ -38,22 +39,38 @@ def __init__(self, auth, condition=None, error_page=None):
     def on_error(self, context):
         if isinstance(context.get("exception"), RefreshError):
             del context["exception"]
-            self.auth.session.clear()
-            if re.search(REGEX_APPJSON, request.headers.get("accept", "")) and (
-                request.headers.get("json-redirects", "") != "on"
-            ):
-                raise HTTP(403)
-            redirect_next = request.fullpath
-            if request.query_string:
-                redirect_next = redirect_next + "?{}".format(request.query_string)
-            self.auth.flash.set("Invalid credentials")
-            redirect(
-                URL(
-                    self.error_page,
-                    vars=dict(next=redirect_next),
-                    use_appname=self.auth.param.use_appname_in_redirects,
-                )
+            self._handle_error()
+            
+    def _handle_error(self):
+        # Removes this Google cookie, trying to enforce loggin in again. 
+        response.delete_cookie("G_ENABLED_IDPS")
+        self.auth.session.clear()
+        if re.search(REGEX_APPJSON, request.headers.get("accept", "")) and (
+            request.headers.get("json-redirects", "") != "on"
+        ):
+            raise HTTP(403)
+        redirect_next = request.fullpath
+        if request.query_string:
+            redirect_next = redirect_next + "?{}".format(request.query_string)
+        self.auth.flash.set("Invalid credentials")
+        redirect(
+            URL(
+                self.error_page,
+                vars=dict(next=redirect_next),
+                use_appname=self.auth.param.use_appname_in_redirects,
             )
+        )
+            
+    def on_request(self, context):
+        super().on_request(context)
+        user = self.auth.session.get("user")
```

Comment:
would not rely on user["email"]. instead assume you have self.auth.user_id and user_info = self.db.auth_credentials[self.auth.user_id] if self.auth.user_id else {}

Reply:
Do not merge it. I will submit a better one.

Reply:
The fix to launch.json os good. But the fix to the scoped login is not a good idea. It worked better before. 

The fundamental problem with Google scoped login is that Google often does not include a refresh token on the callback, and so the credentials to access whatever scopes were requested often expire requiring a new login. 

Checking for the presence of a refresh token is a bad idea since sometimes Google never includes one. 

More generally I am undecided whether the Google scoped login code should be removed from py4web. It does work and I use it in notebookgrader, but perhaps the system is not stable enough and should not be part of the py4web codebase. @mdipierro please advise.

Reply:
I can either leave Google scoped login as it was before, or I can remove it.